src/envs/playroom.py

```diff
@@ -0,0 +1,845 @@
+        next_state.set(block, "pose_y", y)
+        next_state.set(block, "pose_z", z)
+        next_state.set(block, "held", 0.0)
+        next_state.set(block, "clear", 1.0)
+        next_state.set(self._robot, "fingers", fingers)
+        return next_state
+
+    def _transition_stack(self, state: State, action: Action) -> State:
+        next_state = state.copy()
+        # Can only stack if fingers are closed
+        if state.get(self._robot, "fingers") >= self.open_fingers:
+            return next_state
+        # Check that both blocks exist
+        block = self._get_held_block(state)
+        assert block is not None
+        x, y, z, fingers = action.arr
+        other_block = self._get_highest_block_below(state, x, y, z)
+        if other_block is None:  # no block to stack onto
+            return next_state
+        # Can't stack onto yourself!
+        if block == other_block:
+            return next_state
+        # Need block we're stacking onto to be clear
+        if state.get(other_block, "clear") < self.clear_tol:
+            return next_state
+        # Execute stack by snapping into place
+        cur_x = state.get(other_block, "pose_x")
+        cur_y = state.get(other_block, "pose_y")
+        cur_z = state.get(other_block, "pose_z")
+        next_state.set(block, "pose_x", cur_x)
+        next_state.set(block, "pose_y", cur_y)
+        next_state.set(block, "pose_z", cur_z+CFG.playroom_block_size)
+        next_state.set(block, "held", 0.0)
+        next_state.set(block, "clear", 1.0)
+        next_state.set(other_block, "clear", 0.0)
+        next_state.set(self._robot, "fingers", fingers)
+        return next_state
+
+    def _transition_move(self, state: State, action: Action) -> State:
+        x, y, _, rotation, fingers = action.arr
+        next_state = state.copy()
+        next_state.set(self._robot, "pose_x", x)
+        next_state.set(self._robot, "pose_y", y)
+        next_state.set(self._robot, "rotation", rotation)
+        next_state.set(self._robot, "fingers", fingers)
+        return next_state
+
+    def _transition_door(self, state: State, action: Action) -> State:
+        # opens/closes a door that the robot is next to and facing
+        door = self._get_door_next_to(state, action)
+        next_state = state.copy()
+        if state.get(door, "open") < 0.5:
+            next_state.set(door, "open", 1.0)
+        else:
+            next_state.set(door, "open", 0.0)
+        return next_state
+
+    def get_train_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_train_tasks,
+                               rng=self._train_rng)
+
+    def get_test_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_test_tasks,
+                               rng=self._test_rng)
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {self._On, self._OnTable, self._GripperOpen, self._Holding,
+                self._Clear}
```

Comment:
Remember to update all these property methods when you add more predicates/options